azure-ipam/ipam.go

```diff
@@ -0,0 +1,163 @@
+package main
+
+import (
+	"context"
+
+	"github.com/Azure/azure-container-networking/azure-ipam/internal/buildinfo"
+	"github.com/Azure/azure-container-networking/cns"
+	"github.com/Azure/azure-container-networking/network"
+	cniSkel "github.com/containernetworking/cni/pkg/skel"
+	cniTypes "github.com/containernetworking/cni/pkg/types"
+	types100 "github.com/containernetworking/cni/pkg/types/100"
+	"github.com/pkg/errors"
+	"go.uber.org/zap"
+)
+
+// IPAMPlugin is the struct for the delegated azure-ipam plugin
+type IPAMPlugin struct {
+	Name      string
+	Version   string
+	Options   map[string]interface{}
+	logger    *zap.Logger
+	cnsClient cnsClient
+}
+
+type cnsClient interface {
```

Comment:
I don't like the idea of interfaces being used across packages (and here, modules). I prefer to scope the interface to the usage (and declare it there). That makes it easy to mock or swap impls, since then my code is only using the interface that I have defined around just the behavior that I use.

Reply:
fair points 👍

Reply:
^